File: src/main/java/de/maxhenkel/pipez/blocks/GasPipeBlock.java
package de.maxhenkel.pipez.blocks;

import de.maxhenkel.pipez.blocks.tileentity.GasPipeTileEntity;
import de.maxhenkel.pipez.gui.ExtractContainer;
import de.maxhenkel.pipez.gui.containerfactory.PipeContainerProvider;
import de.maxhenkel.pipez.utils.GasUtils;
import de.maxhenkel.pipez.utils.MekanismUtils;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.LevelAccessor;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.phys.BlockHitResult;

public class GasPipeBlock extends PipeBlock {

    protected GasPipeBlock() {
    }

    @Override
    public boolean canConnectTo(Level world, BlockPos pos, Direction facing) {
        if (!MekanismUtils.isMekanismInstalled()) {
            return false;
        }
        return GasUtils.hasChemicalCapability(world, pos.relative(facing), facing.getOpposite());
    }

    @Override
    public boolean isPipe(LevelAccessor world, BlockPos pos, Direction facing) {
        BlockState state = world.getBlockState(pos.relative(facing));
        return state.getBlock().equals(this);
    }

    @Override
    BlockEntity createTileEntity(BlockPos pos, BlockState state) {
        if (!MekanismUtils.isMekanismInstalled()) {
            return null;
        }
        return new GasPipeTileEntity(pos, state);
    }

    @Override
    public InteractionResult onPipeSideActivated(BlockState state, Level worldIn, BlockPos pos, Player player, InteractionHand handIn, BlockHitResult hit, Direction direction) {
        BlockEntity tileEntity = worldIn.getBlockEntity(pos);
        if (tileEntity instanceof GasPipeTileEntity && isExtracting(worldIn, pos, direction)) {
            if (worldIn.isClientSide) {
                return InteractionResult.SUCCESS;
            }
            GasPipeTileEntity pipe = (GasPipeTileEntity) tileEntity;
            PipeContainerProvider.openGui(player, pipe, direction, -1, (i, playerInventory, playerEntity) -> new ExtractContainer(i, playerInventory, pipe, direction, -1));
            return InteractionResult.SUCCESS;
        }
        return super.onPipeSideActivated(state, worldIn, pos, player, handIn, hit, direction);
    }
}

File: src/main/java/de/maxhenkel/pipez/blocks/GasPipeBlock.kt
package de.maxhenkel.pipez.blocks

import de.maxhenkel.pipez.Main
import de.maxhenkel.pipez.blocks.tileentity.GasPipeTileEntity
import de.maxhenkel.pipez.gui.ExtractContainer
import de.maxhenkel.pipez.gui.containerfactory.PipeContainerProvider
import de.maxhenkel.pipez.types.CacheMode
import de.maxhenkel.pipez.connections.CapabilityCache
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.InteractionHand
import net.minecraft.world.InteractionResult
import net.minecraft.world.entity.player.Player
import net.minecraft.world.level.Level
import net.minecraft.world.level.LevelAccessor
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.phys.BlockHitResult

class GasPipeBlock : PipeBlock() {
    init {
        registryName = ResourceLocation(Main.MODID, "gas_pipe")
    }

    override fun canConnectTo(world: LevelAccessor, pos: BlockPos, facing: Direction): Boolean {
        val level = world as? Level ?: return false

        return CapabilityCache.INSTANCE.getGasCapability(
            level = level,
            blockPos = pos.relative(facing),
            direction = facing.opposite,
            cacheMode = CacheMode.REFRESH,
        ).isPresent
    }

    override fun isPipe(world: LevelAccessor, pos: BlockPos, facing: Direction): Boolean {
        val state = world.getBlockState(pos.relative(facing))
        return state.block == this
    }

    override fun createTileEntity(pos: BlockPos, state: BlockState): BlockEntity {
        return GasPipeTileEntity(pos, state)
    }

    override fun onPipeSideActivated(
        state: BlockState,
        worldIn: Level,
        pos: BlockPos,
        player: Player,
        handIn: InteractionHand,
        hit: BlockHitResult,
        direction: Direction
    ): InteractionResult {
        val tileEntity = worldIn.getBlockEntity(pos)
        if (tileEntity is GasPipeTileEntity && isExtracting(worldIn, pos, direction)) {
            if (worldIn.isClientSide) {
                return InteractionResult.SUCCESS
            }
            PipeContainerProvider.openGui(player, tileEntity, direction, -1) { i, playerInventory, _ ->
                ExtractContainer(i, playerInventory, tileEntity, direction, -1)
            }
            return InteractionResult.SUCCESS
        }
        return super.onPipeSideActivated(state, worldIn, pos, player, handIn, hit, direction)
    }
}